### PR TITLE
Constant values should appear on the right-hand side of comparison

### LINKED ISSUE #40561: Is/should there a code-style for avoiding null on left side of a binary expression?
Is there are code-style for avoiding null on left side of a binary expression? If not an explicit requirement, maybe a guideline?
https://github.com/dotnet/runtime/blob/master/docs/coding-guidelines/coding-style.md
https://github.com/JosefPihrt/Roslynator/blob/master/docs/analyzers/RCS1098.md

### PR DESCRIPTION
**Pull request by elachlan**
Fixes #40561

### CODE REVIEW

File: src/coreclr/src/System.Private.CoreLib/src/System/GC.cs [hunk 1]
```diff
@@ -133,7 +133,7 @@ public static void AddMemoryPressure(long bytesAllocated)
                         SR.ArgumentOutOfRange_NeedPosNum);
             }
 
-            if ((4 == IntPtr.Size) && (bytesAllocated > int.MaxValue))
+            if ((IntPtr.Size == 4) && (bytesAllocated > int.MaxValue))
```
Comment (Clockwork-Muse) [inline src/coreclr/src/System.Private.CoreLib/src/System/GC.cs:136]: `IntPtr.Size == sizeof(int)`?

File: src/libraries/Common/src/System/Data/Common/MultipartIdentifier.cs [hunk 2]
```diff
@@ -68,7 +68,7 @@ private static bool IsWhitespace(char ch)
                 throw ADP.InvalidMultipartNameToManyParts(property, name, limit);
             }
 
-            if (-1 != leftQuote.IndexOf(separator) || -1 != rightQuote.IndexOf(separator) || leftQuote.Length != rightQuote.Length)
+            if (leftQuote.IndexOf(separator) != -1 || rightQuote.IndexOf(separator) != -1 || leftQuote.Length != rightQuote.Length)
```
Comment (Clockwork-Muse) [inline src/libraries/Common/src/System/Data/Common/MultipartIdentifier.cs:71]: .... should we have a `NOT_FOUND` constant for the `IndexOf` comparisons?

Also, the current comparisons make my brain hurt for what they're saying ("not not found").  Should some of these be switched to `Contains`, and others to testing against a positive index?  (In another PR, obviously)

File: src/coreclr/src/System.Private.CoreLib/src/System/Text/StringBuilder.CoreCLR.cs [hunk 3]
```diff
@@ -145,7 +145,7 @@ internal unsafe void InternalCopy(IntPtr dest, int len)
         /// <param name="byteIndex">The logical byte index in this builder.</param>
         private StringBuilder FindChunkForByte(int byteIndex)
         {
-            Debug.Assert(0 <= byteIndex && byteIndex <= Length * sizeof(char));
```
Comment (Clockwork-Muse) [inline src/coreclr/src/System.Private.CoreLib/src/System/Text/StringBuilder.CoreCLR.cs:148]: This (and a bunch of other places) seems to be more likely to have been done in the style of range checks, than the normal assignment concerns.  Assuming we're changing it anyways.  Would be nice to have a better range check syntax, though.

File: src/libraries/System.Data.Common/src/System/Data/Common/AdapterUtil.Common.cs [hunk 4]
```diff
@@ -785,10 +785,10 @@ internal static void BuildSchemaTableInfoTableNames(string[] columnNameArray)
             Dictionary<string, int> hash = new Dictionary<string, int>(columnNameArray.Length);
 
             int startIndex = columnNameArray.Length; // lowest non-unique index
-            for (int i = columnNameArray.Length - 1; 0 <= i; --i)
+            for (int i = columnNameArray.Length - 1; i >= 0; --i)
             {
                 string columnName = columnNameArray[i];
-                if ((null != columnName) && (0 < columnName.Length))
+                if ((columnName != null) && (columnName.Length > 0))
```
Comment (Clockwork-Muse) [inline src/libraries/System.Data.Common/src/System/Data/Common/AdapterUtil.Common.cs:791]: `!String.IsNullOrEmpty(columnName)`?

File: src/libraries/System.Data.Common/src/System/Data/Common/DataStorage.cs [hunk 5]
```diff
@@ -560,9 +560,9 @@ protected void SetNullStorage(BitArray nullbits)
         internal static Type GetType(string value)
         {
             Type? dataType = Type.GetType(value); // throwOnError=false, ignoreCase=fase
-            if (null == dataType)
+            if (dataType == null)
             {
-                if ("System.Numerics.BigInteger" == value)
+                if (value == "System.Numerics.BigInteger")
```
Comment (Clockwork-Muse) [inline src/libraries/System.Data.Common/src/System/Data/Common/DataStorage.cs:565]: `value == typeof(System.Numerics.BigInteger).FullName`?  Don't know all the implications of that though.

File: src/coreclr/src/System.Private.CoreLib/src/System/GC.cs (hunk 1 above)
Comment (elachlan) [reply src/coreclr/src/System.Private.CoreLib/src/System/GC.cs:136]: Isn't this a 32bit check? Don't we have a utility function to return if the system is 32bit? msbuild does something similar:
https://github.com/dotnet/msbuild/blob/master/src/Shared/EnvironmentUtilities.cs#L11

Comment (elachlan) [reply src/coreclr/src/System.Private.CoreLib/src/System/GC.cs:136]: No, it is a bitness check for 32bit.
https://github.com/dotnet/runtime/blob/ea2b09beea535abf960651c59359e40f99f2fecf/src/libraries/System.Private.CoreLib/src/System/Environment.cs#L138

File: src/libraries/System.Data.Common/src/System/Data/Common/DataStorage.cs (hunk 5 above)
Comment (elachlan) [reply src/libraries/System.Data.Common/src/System/Data/Common/DataStorage.cs:565]: There is a comment above the function:
```
/// Types like "System.Guid" will load regardless of AssemblyQualifiedName because they are special
/// Types like "System.Data.SqlTypes.SqlString" will load because they are in the same assembly as this code
/// Types like "System.Numerics.BigInteger" won't load because they are not special and not same assembly as this code
```

Comment (elachlan) [reply src/libraries/System.Data.Common/src/System/Data/Common/DataStorage.cs:565]: I don't think that means we have to use the string, but I am not 100% sure.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/TraceLogging/Statics.cs [hunk 6]
```diff
@@ -169,7 +169,7 @@ public static void EncodeTags(int tags, ref int pos, byte[]? metadata)
 
         public static void CheckName(string? name)
         {
-            if (name != null && 0 <= name.IndexOf('\0'))
+            if (name != null && name.Contains('\0'))
```
Comment (elachlan) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/TraceLogging/Statics.cs:172]: This is causing me issues and I am not sure how to fix it so that it compiles.

File: src/coreclr/src/System.Private.CoreLib/src/System/GC.cs (hunk 1 above)
Comment (Clockwork-Muse) [reply src/coreclr/src/System.Private.CoreLib/src/System/GC.cs:136]: .... I'm a little surprised we don't have a 32-bit process check...  But it's probably from the changeover 32->64, so code was going to be written as `if(Is64BitProcess) {...}`...

Yes, I'm aware it's a 32-bit check; the use of `sizeof` was just an attempt to avoid a magic number.

File: src/libraries/System.Data.Common/src/System/Data/Common/DataStorage.cs (hunk 5 above)
Comment (Clockwork-Muse) [reply src/libraries/System.Data.Common/src/System/Data/Common/DataStorage.cs]: `typeof` has to work, because it already does it the line below anyways.

File: src/coreclr/src/System.Private.CoreLib/src/System/GC.cs (hunk 1 above)
Comment (elachlan) [reply src/coreclr/src/System.Private.CoreLib/src/System/GC.cs:136]: I think using `sizeof` would only run it when the process bitness and the os bitness match. Which I don't think is the intention of the code.

Comment (stephentoub) [reply src/coreclr/src/System.Private.CoreLib/src/System/GC.cs:136]: This PR is changing ~600 files.  Please restrict the change to be purely switching the sides of the expression and nothing else.  It's way too likely mistakes will otherwise creep in.

**Comment by danmosemsft** [comment]
I'm going to mark "no-merge" for now, since this should not be merged until master is a 6.0 branch, about a week out.

File: src/libraries/Common/src/System/Collections/Generic/EnumerableHelpers.cs [hunk 7]
```diff
@@ -70,7 +70,7 @@ internal static T[] ToArray<T>(IEnumerable<T> source, out int length)
                                 int newLength = count << 1;
                                 if ((uint)newLength > MaxArrayLength)
                                 {
-                                    newLength = MaxArrayLength <= count ? count + 1 : MaxArrayLength;
+                                    newLength = count >= MaxArrayLength ? count + 1 : MaxArrayLength;
```
Comment (danmosemsft) [inline src/libraries/Common/src/System/Collections/Generic/EnumerableHelpers.cs:73]: I thought this change was only about equality operators. I am not sure I see the value of reordering inequality operators.

File: src/coreclr/src/System.Private.CoreLib/src/System/GC.cs (hunk 1 above)
Comment (danmosemsft) [reply src/coreclr/src/System.Private.CoreLib/src/System/GC.cs:136]: You might also consider breaking it into pieces -- doing a few folders in one PR, merging that, and doing another PR, and enabling the rule in the last one. Or just doing 20 typical changes in one PR. That would also help us decide whether we're making the right change.

File: src/libraries/System.Data.Common/src/System/Data/Common/DataStorage.cs (hunk 5 above)
Comment (danmosemsft) [reply src/libraries/System.Data.Common/src/System/Data/Common/DataStorage.cs]: Please let's limit this change to just reordering equality operators. Future changes are always possible, but making the PR(s) as simple and mechanical as possible makes it easier to review and less likely to introduce a bug.

Comment (elachlan) [reply src/libraries/System.Data.Common/src/System/Data/Common/DataStorage.cs]: @danmosemsft are you okay if we create a separate PR for this? Or will we need an issue first?

File: src/coreclr/src/System.Private.CoreLib/src/System/GC.cs (hunk 1 above)
Comment (elachlan) [reply src/coreclr/src/System.Private.CoreLib/src/System/GC.cs:136]: I'd like to do one big change so we see how much is affected. Once we know how much of the code-base is touched then I can split it up with a PR per library or something.

At the moment I am having issues refactoring due to a nuget issue related to I think #32205.

File: src/libraries/Common/src/System/Collections/Generic/EnumerableHelpers.cs (hunk 7 above)
Comment (elachlan) [reply src/libraries/Common/src/System/Collections/Generic/EnumerableHelpers.cs:73]: SA1131 was flagged on this part of the code, so it was changed.

File: src/coreclr/src/System.Private.CoreLib/src/System/GC.cs (hunk 1 above)
Comment (Clockwork-Muse) [reply src/coreclr/src/System.Private.CoreLib/src/System/GC.cs:136]: `sizeof(<primitive>)` returns a constant value at compile time.  Unlike C++, primitives have a fixed size in C# (which is why there's a request for `nint` - Native INTeger).

File: src/libraries/System.Data.Common/src/System/Data/Common/DataStorage.cs (hunk 5 above)
Comment (danmosemsft) [reply src/libraries/System.Data.Common/src/System/Data/Common/DataStorage.cs]: > @danmosemsft are you okay if we create a separate PR for this? Or will we need an issue first?

It's a judgement call, if something is a small change, and you're pretty sure it's a good one, sometimes people throw up a PR without an issue. We don't have a hard rule like some communities. But an issue is generally a good idea, as it avoids wasting time making a code change we wouldn't take or that would need reworking to a new approach.

File: src/libraries/Common/src/System/Collections/Generic/EnumerableHelpers.cs (hunk 7 above)
Comment (danmosemsft) [reply src/libraries/Common/src/System/Collections/Generic/EnumerableHelpers.cs:73]: OK, fair enough. I guess my suggestion is as above then, to bite off a smaller part of the tree, and get it clean for SA1131. You could force-push that smaller change to this PR maybe.

File: src/libraries/Common/src/System/Security/Cryptography/Asn1/AlgorithmIdentifierAsn.xml.cs [hunk 8]
```diff
@@ -1,4 +1,4 @@
-// Licensed to the .NET Foundation under one or more agreements.
+﻿// Licensed to the .NET Foundation under one or more agreements.
```
Comment (Clockwork-Muse) [inline src/libraries/Common/src/System/Security/Cryptography/Asn1/AlgorithmIdentifierAsn.xml.cs:1]: Whitespace is the only difference in this file.  I'm all for cleaning up, but should that be a separate PR (ie, as part of an autoformatter)?

File: src/libraries/Common/src/System/Security/Cryptography/Asn1/ECDomainParameters.xml.cs [hunk 9]
```diff
@@ -1,4 +1,4 @@
-// Licensed to the .NET Foundation under one or more agreements.
+﻿// Licensed to the .NET Foundation under one or more agreements.
```
Comment (Clockwork-Muse) [inline src/libraries/Common/src/System/Security/Cryptography/Asn1/ECDomainParameters.xml.cs:1]: Multiple of these

File: src/coreclr/src/System.Private.CoreLib/src/System/Text/StringBuilder.CoreCLR.cs (hunk 3 above)
Comment (danmosemsft) [reply src/coreclr/src/System.Private.CoreLib/src/System/Text/StringBuilder.CoreCLR.cs:148]: Yeah, it makes me slightly sad to lose this pattern. because I find it easier to quickly see it's correct. Another example
```
<             Debug.Assert(0 <= startPos && startPos <= s.Length);
>             Debug.Assert(startPos >= 0 && startPos <= s.Length);
```

**Comment by danmosemsft** [comment]
@elachlan did you make all the changes mechanically with the analyzer - literally just setting it to Warning and clearing all the warnings? Does it do the fixes for you? If you made any fixups is there a way to keep them in a separate commit in the PR, so those can be checked more closely?

Please revert all the edits to the license headers (line endings maybe)?

File: src/coreclr/src/System.Private.CoreLib/src/System/Text/StringBuilder.CoreCLR.cs (hunk 3 above)
Comment (danmosemsft) [reply src/coreclr/src/System.Private.CoreLib/src/System/Text/StringBuilder.CoreCLR.cs:148]: Another random one
```
-            Debug.Assert(0 <= charsConsumedSoFar && charsConsumedSoFar <= originalCharCount, "Caller should've checked this condition.");
+            Debug.Assert(charsConsumedSoFar >= 0 && charsConsumedSoFar <= originalCharCount, "Caller should've checked this condition.");
```

Comment (elachlan) [reply src/coreclr/src/System.Private.CoreLib/src/System/Text/StringBuilder.CoreCLR.cs:148]: Maybe we can get stylecop to change its detection to ignore range checks?

Someone flagged it here:
https://github.com/DotNetAnalyzers/StyleCopAnalyzers/issues/2830

**Comment by danmosemsft** [comment]
I scanned through the 4,713 lines edited and it seems OK to me aside from comments above. I will feel better if you can help me identify which changes specifically were not done by the analyzer itself, by segregating into a commit. Taht way in theory I can run the analyzer myself and reproduce your original commit, then just examine any hand fixups you did.

File: src/coreclr/src/System.Private.CoreLib/src/System/Text/StringBuilder.CoreCLR.cs (hunk 3 above)
Comment (danmosemsft) [reply src/coreclr/src/System.Private.CoreLib/src/System/Text/StringBuilder.CoreCLR.cs:148]: I doubt it - that would upset the other camp 😉 I think we can live with it, since we mix both already.

**Comment by elachlan** [comment]
> I scanned through the 4,713 lines edited and it seems OK to me aside from comments above. I will feel better if you can help me identify which changes specifically were not done by the analyzer itself, by segregating into a commit. Taht way in theory I can run the analyzer myself and reproduce your original commit, then just examine any hand fixups you did.

The only time I made manual changes was when the project would not compile due to environment issues (duplicate framework issue) or changes from review. Manual changes were VERY few and could have been done using the analyzers if the project was compiling okay. There were a couple of instances where the refactor caused another analyzer to be triggered causing a compile error. So those were fixed using the appropriate analyzer.

The main manual change was to the CodeAnalysis.ruleset

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/TraceLogging/Statics.cs (hunk 6 above)
Comment (elachlan) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/TraceLogging/Statics.cs:172]: @danmosemsft this is what is causing the build failure. If we use IndexOf then the code is marked by one of the analyzers with a warning, causing a build failure. If we use contains, then the net48 build fails. Since net48 doesn't have an overloads for char.

Comment (danmosemsft) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/TraceLogging/Statics.cs:172]: These tracing files are built into 2 projects, one is corelib and the other is a redist project for .NET Framework use. This second one is causing the problem here. 
https://dev.azure.com/dnceng/public/_build/results?buildId=779866&view=logs&j=ed31f2e8-9a21-528b-426c-e91b77680964&t=0cc22ee9-e6e6-58fc-cb15-c6bbf3030f86&l=930 
the fix is to check for ES_BUILD_STANDALONE and if it's set then use the .NET Framework version
eg 
https://github.com/danmosemsft/runtime/blob/9646df5d76213e673c2cb44aced17113a4961a01/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs#L3797

**Comment by danmosemsft** [comment]
@elachlan could you please resolve conflicts, then I will review again? 

If there are any remaining unrelated changes that you can revert to reduce the diff, that would be helpful. I'm prepared to read all 4700 lines again though.

**Comment by elachlan** [comment]
@danmosemsft I think it might be best if I close this PR and a new one is created based off the latest. Would you prefer I target a single project at a time per PR? with a final PR for enabling the analyzer?

**Comment by danmosemsft** [comment]
@elachlan that sounds like a good idea. Could you please also submit smaller PR's, as they have a much easier time getting reviewed and merged without conflicts appearing.

For example, could you start with just a PR for src/libraries/System.Private.CoreLib ?
